src/jace/jax/translation_cache.py

```diff
@@ -0,0 +1,297 @@
+from abc import abstractmethod
+from collections import OrderedDict
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Final, Protocol, TypeAlias, runtime_checkable
+
+import dace
+from jax import core as jax_core
+
+from jace import util
+
+
+if TYPE_CHECKING:
+    from jace.jax import stages
+
+# This is the default cache size we are using
+_DEF_CACHE_SIZE: Final[int] = 256
+
+# This are the caches that we are using.
+_TRANSLATION_CACHES: dict[type[stages.Stage], TranslationCache] = {}
+
+
+def cached_translation(
+    action: Callable,
+) -> Callable:
+    """Decorator for making the transfer method, i.e. `JaceWrapped.lower()` and `JaceLowered.compile()` cacheable.
+
+    The main issue is that we can not simply cache on the actual arguments we pass to them, but on an abstract
+    (or concrete; static arguments + compiling) description on them, and this is what this decorator is for.
+    Based on its argument it will generate a key of the call, see `TranslationCache.make_key()` for more.
+    Then it will check if the result is known and if needed it will perform the actual call.
+
+    Beside this the function will two two things.
+    The first is, that it will set the `_cache` member of `self` to the associated cache.
+    Thus an annotated object need to define such a member.
+
+    The second thing it will do is optional, if the call is not cached inside the cache the wrapped function has to be run.
+    In that case the wrapper will first check if the object defines the `_call_description` member.
+    If this is the case the wrapper will set this object to an abstract description of the call, which is also used as key in the cache.
+    After the function return this member is set to `None`.
+    """
+
+    @ft.wraps(action)
+    def _action_wrapper(
+        self: stages.JaceWrapped | stages.JaceLowered,
+        *args: Any,
+        **kwargs: Any,
+    ) -> stages.Stage:
+        # Get the abstract description of the call, that is used as key.
+        key: CachedCallDescription = self._make_call_decscription(*args, **kwargs)
+        if self._cache.has(key):
+            return self._cache.get(key)
+
+        # We must actually perform the call
+        next_stage: stages.Stage = action(self, *args, **kwargs)
+        self._cache.add(key, next_stage)
+        return next_stage
+
+    return _action_wrapper
+
+
+def clear_translation_cache() -> None:
+    """Clear all caches associated to translation."""
+    _TRANSLATION_CACHES.clear()
+
+
+def get_cache(
+    stage: stages.Stage,
+) -> TranslationCache:
+    """Returns the cache that is used for `stage`."""
+    # The caches are per stage and not per instance basis
+    tstage = type(stage)
+    if tstage not in _TRANSLATION_CACHES:
+        _TRANSLATION_CACHES[tstage] = TranslationCache(size=_DEF_CACHE_SIZE)
+    return _TRANSLATION_CACHES[tstage]
+
+
+@dataclass(init=True, eq=True, frozen=True)
```

Comment:
I would remove the arguments matching the default values.
```suggestion
@dataclass(frozen=True)
```